docker/docker.go

```diff
@@ -22,201 +22,460 @@ limitations under the License.
-	groupWrap      map[string]wrapper.Stats     // wrapper for cgroup name and interface for stats extraction
-	hostname       string                       // name of the host
+	containers map[string]containerData      // holds data for a container under its short id
+	client     client.DockerClientInterface  // client for communication with docker (basic info, stats, mount points)
+	list       map[string]dock.APIContainers // contain list of all available docker containers with info about their specification
+
+}
+
+type dynamicElement struct {
+	name        string
+	description string
+}
+
+// dynamic metrics holds items of dynamic element in metric namespace which occurs after the key-word
+var dynamicMetrics = map[string]dynamicElement{
+	"filesystem":   dynamicElement{"device_name", "a name of filesystem device"},
+	"labels":       dynamicElement{"label", "a name of container label"},
+	"network":      dynamicElement{"network_interface", "a name of network interface or 'total' for aggregate"},
+	"percpu_usage": dynamicElement{"cpu_id", "an id of cpu"},
 }
 
 // Docker plugin initializer
-func NewDocker() (*docker, error) {
-	host, _ := os.Hostname()
-	// create new docker client
-	dockerClient := client.NewDockerClient()
-	// list all running containers
-	containers, err := dockerClient.ListContainers()
+func New() *docker {
+	return &docker{
+		containers: map[string]containerData{},
+		client:     client.NewDockerClient(),
+		list:       map[string]dock.APIContainers{},
+	}
+}
+
+// CollectMetrics retrieves values of requested metrics
+func (d *docker) CollectMetrics(mts []plugin.MetricType) ([]plugin.MetricType, error) {
+	var err error
+	metrics := []plugin.MetricType{}
+	d.list = map[string]dock.APIContainers{}
+
+	// get list of possible network metrics
+	networkMetrics := []string{}
+	utils.FromCompositionTags(wrapper.NetworkInterface{}, "", &networkMetrics)
 
+	// get list of all running containers
+	d.list, err = d.client.ListContainersAsMap()
 	if err != nil {
+		fmt.Fprintln(os.Stderr, "The list of running containers cannot be retrived, err=", err)
 		return nil, err
 	}
 
-	d := &docker{
-		stats:          cgroups.NewStats(),
-		client:         dockerClient,
-		tools:          new(tls.MyTools),
-		containersInfo: containers,
-		groupWrap:      wrapper.Cgroups2Stats,
-		hostname:       host}
-
-	return d, nil
-}
+	// retrieve requested docker ids
+	rids, err := d.getRequestedIDs(mts...)
+	if err != nil {
+		return nil, err
+	}
 
-// wrapper for cgroup stats extraction
-func (d *docker) getStats(id string) error {
+	// for each requested id set adequate item into docker.container struct with stats
+	for _, rid := range rids {
 
-	for cg, stat := range d.groupWrap {
-		// find mount point for each cgroup
-		mp, err := d.client.FindCgroupMountpoint(cg)
+		if contSpec, exist := d.list[rid]; !exist {
+			return nil, fmt.Errorf("Docker container does not exist, container_id=%s", rid)
+		} else {
```

Comment:
`else` block is necessary because `contSpec` value is only available inside 'else' block.